api/server.go

```diff
@@ -0,0 +1,125 @@
+package api
+
+import (
+	"context"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/improbable-eng/grpc-web/go/grpcweb"
+	ds "github.com/ipfs/go-datastore"
+	logging "github.com/ipfs/go-log"
+	tserv "github.com/textileio/go-textile-core/threadservice"
+	pb "github.com/textileio/go-textile-threads/api/pb"
+	es "github.com/textileio/go-textile-threads/eventstore"
+	"github.com/textileio/go-textile-threads/util"
+	logger "github.com/whyrusleeping/go-logging"
+	"google.golang.org/grpc"
+)
+
+var (
+	log = logging.Logger("api")
+)
+
+type Server struct {
+	rpc     *grpc.Server
+	proxy   *http.Server
+	service *service
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+type Options struct {
+	Addr      string // defaults to 0.0.0.0:9090
+	ProxyAddr string // defaults to 0.0.0.0:9091
+	Debug     bool
+}
+
+func NewServer(ctx context.Context, ds ds.Datastore, ts tserv.Threadservice, opts Options) (*Server, error) {
+	var err error
+	if opts.Debug {
+		err = util.SetLogLevels(map[string]logger.Level{
+			"api": logger.DEBUG,
+		})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	s := &Server{
+		rpc: grpc.NewServer(),
+		service: &service{
+			datastore: ds,
+			threads:   ts,
+			stores:    make(map[string]*es.Store), // @todo how to reload stores on start?
+		},
+		ctx:    ctx,
+		cancel: cancel,
+	}
+
+	if opts.Addr == "" {
+		opts.Addr = "0.0.0.0:9090"
+	}
+	listener, err := net.Listen("tcp", opts.Addr)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		pb.RegisterAPIServer(s.rpc, s.service)
+		s.rpc.Serve(listener)
+	}()
+
+	webrpc := grpcweb.WrapServer(
+		s.rpc,
+		grpcweb.WithOriginFunc(func(origin string) bool {
+			return true
+		}),
+		grpcweb.WithWebsockets(true),
+		grpcweb.WithWebsocketOriginFunc(func(req *http.Request) bool {
+			return true
+		}))
+	if opts.ProxyAddr == "" {
+		opts.ProxyAddr = "0.0.0.0:9091"
+	}
+	s.proxy = &http.Server{
+		Addr: opts.ProxyAddr,
+	}
+	s.proxy.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if webrpc.IsGrpcWebRequest(r) ||
+			webrpc.IsAcceptableGrpcCorsRequest(r) ||
+			webrpc.IsGrpcWebSocketRequest(r) {
+			webrpc.ServeHTTP(w, r)
+		}
+	})
+
+	errc := make(chan error)
+	go func() {
+		errc <- s.proxy.ListenAndServe()
+		close(errc)
+	}()
+	go func() {
+		for err := range errc {
+			if err != nil && err != http.ErrServerClosed {
+				log.Errorf("proxy error: %s", err)
+			}
+		}
```

Comment:
Shouldn't break the `for` loop if `err == http.ErrServerClosed`?

Reply:
Oop, copy pasta from `threads`, probably should change that too.